File: packages/backend/lib/params.py
from typing import Literal, TypedDict


class SelectParam(TypedDict):
    type: Literal["select"]
    options: list[str]


class IntParam(TypedDict):
    type: Literal["int"]
    suffix: str | None
    value: int


class FloatParam(TypedDict):
    type: Literal["float"]
    suffix: str | None
    value: float


class StrParam(TypedDict):
    type: Literal["str"]
    value: str


class BoolParam(TypedDict):
    type: Literal["bool"]
    value: bool


type AllParamTypes = (
    SelectParam | IntParam | FloatParam | StrParam | BoolParam | CompositeParam
)


class CompositeParam(TypedDict):
    type: Literal["composite"]
    children: dict[str, "AllParamTypes"]

File: packages/shared/types/params.ts

export interface SelectStrParam {
    type: "select.str"
    options: string[]
    value: string
}

export interface SelectIntParam {
    type: "select.int"
    options: number[]
    value: number
}

export interface SelectFloatParam {
    type: "select.float"
    options: number[]
    value: number
}

export interface IntParam {
    type: "int"
    suffix?: string
    value: number
}


export interface FloatParam {
    type: "float"
    suffix?: string
    value: number
}


export interface StrParam {
    type: "str"
    value: string
}

export interface BoolParam {
    type: "bool"
    value: boolean
}

export interface InstanceEquipmentParam {
    type: "instance.equipment"
    instance_id: string
    instance: string
}

export interface InstanceExperimentParam {
    type: "instance.experiment"
    instance_id: string
    instance: string
}

export type AllParamTypes = (
    SelectStrParam | SelectFloatParam | SelectIntParam | IntParam | FloatParam | StrParam | BoolParam |
    //  CompositeParam |
    InstanceEquipmentParam | InstanceExperimentParam
)


// export interface CompositeParam {
//     type: "composite"
//     children: Record<string, AllParamTypes>
// }